Mathlib/Algebra/DirectSum/Internal.lean

```diff
@@ -337,6 +346,90 @@ theorem DirectSum.coeAlgHom_of [AddMonoid ι] [CommSemiring S] [Semiring R] [Alg
 
 end DirectSum
 
+/-! ### Facts about grade zero -/
+
+namespace SetLike.GradeZero
+
+section Semiring
+variable [Semiring R] [AddMonoid ι] [SetLike σ R] [AddSubmonoidClass σ R]
+variable (A : ι → σ) [SetLike.GradedMonoid A]
+
+/-- The subsemiring `A 0` of `R`. -/
+def subsemiring : Subsemiring R where
+  carrier := A 0
+  __ := submonoid A
+  add_mem' := add_mem
+  zero_mem' := zero_mem (A 0)
+
+/-- The semiring `A 0` inherited from `R` in the presence of `SetLike.GradedMonoid A`. -/
+instance instSemiring : Semiring (A 0) := (subsemiring A).toSemiring
```

Comment:
I'm a bit worried that these instances might be a bad idea; I'm not sure if lean actually has any way to infer `A`, and this might be the explanation for the simpNF failure and the other issues you saw before you moved the imports around.

Perhaps @vierkantor or @digama0 has opinions?

Reply:
I can't reproduce the `simpNF` failure yet:
```
set_option trace.Meta.synthInstance true in
set_option trace.Meta.Tactic.simp true in
example (n : ℕ) : (n : A 0) = (n : R) := by simp only [coe_natCast]
```
succeeds, and the trace shows `  [] ✅ apply @SetLike.GradeZero.instSemiring to Semiring (↥A 0) ▶`. But yeah, it could very well turn out that the higher-order matching problem `A 0 = ?A 0` is too difficult.

Reply:
The simpNF linter is still complaining about line 369 of this file even though the imports are now working. Would you be happy with them being local instances? At the end of the day, in algebraic geometry over a field, A 0 is typically that field. 

Perhaps another approach would be to assume that R is a comm ring, A is a graded R-algebra and the map R -> A identifies R with A 0; then one can work with R instead of A 0 and this is probably enough in applications I guess?

Reply:
An easy way out is to write `GradeZero.subring A` instead of `A 0`, which definitely doesn't cause any instance issue.

I'm also happy with merging them as is, and revisiting it if/when we notice issues.

Reply:
> Is there a natural Lean way of saying "I am a graded R-algebra and actually R is precisely my A 0?"

Not that I know of. We're also missing this case for "I am an R-algebra and actually my center is precisely R"

Reply:
I've added some TODO comments warning about the unification performance here, I think it's then safe to merge as is.